### PR TITLE
app-shells/mksh: Fix the live ebuild and bump 57 from it

### PR DESCRIPTION
**Pull request by lanodan**

### CODE REVIEW

File: app-shells/mksh/mksh-57.ebuild [hunk 1]
```diff
@@ -19,13 +29,14 @@ S="${WORKDIR}/${PN}"
...
-	if use static ; then
+	if use static
+	then
 		unset CC
 		export CC="/usr/bin/klcc"
 		export LDSTATIC="-static"
 	fi
 	export CPPFLAGS="${CPPFLAGS} -DMKSH_DEFAULT_PROFILEDIR=\\\"${EPREFIX}/etc\\\""
-	sh Build.sh -r || die
+	sh Build.sh -r -c lto || sh Rebuild.sh || die
```
Comment (juippis) [inline app-shells/mksh/mksh-57.ebuild]: I don't think this works... what are you trying to do here? Do both files exists, or does Build.sh create the 2nd file? Anyway if this logic is correct, try with
`(sh Build.sh -r -c lto || sh Rebuild.sh) || die` instead. This only works with EAPI-7.

Comment (lanodan) [reply app-shells/mksh/mksh-57.ebuild]: It didn’t seems much wrong with a quick glance, and this is from the live ebuild.

Doing your proposed change for EAPI-7.

Comment (lanodan) [reply app-shells/mksh/mksh-57.ebuild]: In the end `Rebuild.sh` looks to have been removed from mksh, I guess I might have looked at a old cvs checkout.

**Comment by lanodan** [comment]
@juippis 
Do you think I should remove `mksh-57.ebuild` as it’s not keyworded to stable?

**Comment by juippis** [comment]
> Do you think I should remove `mksh-57.ebuild` as it’s not keyworded to stable?

Only drop it, if you are sure your changes won't break anything for users. They then have hard time to revert back into previous working version. 

I would keep it for 30 days after revbumping, then if no new bug reports arise (or if they are fixed hastily) remove it later.

I got a bit stuck with what you do in pkg_postinst, haven't seen that before but it does look a bit wild. Wanted to test it before commenting anything, but I'll do that later.

**Comment by lanodan** [comment]
[2019-03-23 06:54:22] juippis:
> > Do you think I should remove `mksh-57.ebuild` as it’s not keyworded to stable?
> 
> Only drop it, if you are sure your changes won't break anything for users. They then have hard time to revert back into previous working version. 
> I would keep it for 30 days after revbumping, then if no new bug reports arise (or if they are fixed hastily) remove it later.

I’ll probably do that since at least `-c lto` could introduce bugs (even if it shouldn’t).

> I got a bit stuck with what you do in pkg_postinst, haven't seen that before but it does look a bit wild. Wanted to test it before commenting anything, but I'll do that later.

It looks like it's (weirdly) not present in zsh and fish even through it should be required to update this file, maybe it's from a eclass or their installer already modifies `/etc/shells`.

It is a bit weird but if I had to rewrite it, it would be more something like:
```
if grep -q -v /bin/mksh "${ROOT}/etc/shells"
then
	ebegin "Adding mksh to /etc/shells"
	echo '/bin/mksh' >> "${ROOT}etc/shells"
	eend $?
fi
```
This way it’s only triggered when mksh isn’t present in `/etc/shells` rather than updating the file at each install.

**Comment by lanodan** [comment]
[2019-03-23 17:51:21+0100] Haelwenn Monnier:
> > I got a bit stuck with what you do in pkg_postinst, haven't seen that before but it does look a bit wild. Wanted to test it before commenting anything, but I'll do that later.
> 
> It looks like it's (weirdly) not present in zsh and fish even through it should be required to update this file, maybe it's from a eclass or their installer already modifies `/etc/shells`.
> 
> It is a bit weird but if I had to rewrite it, it would be more something like:
> ```
> if grep -q -v /bin/mksh "${ROOT}/etc/shells"
> then
> 	ebegin "Adding mksh to /etc/shells"
> 	echo '/bin/mksh' >> "${ROOT}etc/shells"
> 	eend $?
> fi
> ```
> This way it’s only triggered when mksh isn’t present in `/etc/shells` rather than updating the file at each install.

Well, fish and zsh do not have this issue because baselayout already puts them in /et/shells, I guess I should file a bug for baselayout to add mksh and in the meantime have postinst rewrote like I proposed.

**Comment by lanodan** [comment]
[2019-03-31 01:39:26-0700] Michał Górny:
> I'm sorry but I'm against bumping m-needed packages via p-m unless you are going to maintain it from now on.

Oh yeah makes sense, then I can take this package, I'm using mksh as default shell everywhere, I'm open to co-maintainance btw.

File: app-shells/mksh/mksh-9999.ebuild [hunk 2]
```diff
@@ -12,29 +12,31 @@ if [[ $PV = 9999 ]]; then
...
 RDEPEND=""
 S="${WORKDIR}/${PN}"
 
 src_compile() {
 	tc-export CC
 	# we want to build static with klibc
-	if use static; then unset CC; export CC="/usr/bin/klcc"; export LDSTATIC="-static"; fi
+	if use static
+	then
+		unset CC
```
Comment (mgorny) [inline app-shells/mksh/mksh-9999.ebuild]: Unsetting it makes no sense if you redefine it immediately.

File: app-shells/mksh/mksh-9999.ebuild [hunk 3]
```diff
@@ -12,29 +12,31 @@ if [[ $PV = 9999 ]]; then
...
 IUSE="static"
+DEPEND="static? ( dev-libs/klibc )"
 RDEPEND=""
 S="${WORKDIR}/${PN}"
 
 src_compile() {
 	tc-export CC
 	# we want to build static with klibc
-	if use static; then unset CC; export CC="/usr/bin/klcc"; export LDSTATIC="-static"; fi
+	if use static
```
Comment (mgorny) [inline app-shells/mksh/mksh-9999.ebuild]: We generally keep to single-line `if ...; then` in Gentoo.

File: app-shells/mksh/mksh-9999.ebuild [hunk 4]
```diff
@@ -12,29 +12,31 @@ if [[ $PV = 9999 ]]; then
...
+	then
+		unset CC
+		export CC="/usr/bin/klcc"
+		export LDSTATIC="-static"
+	fi
 	export CPPFLAGS="${CPPFLAGS} -DMKSH_DEFAULT_PROFILEDIR=\\\"${EPREFIX}/etc\\\""
-	sh Build.sh -r -c lto || sh Rebuild.sh || die
+	sh Build.sh -r -c lto || die
```
Comment (mgorny) [inline app-shells/mksh/mksh-9999.ebuild]: What does `-c lto` do?

File: app-shells/mksh/mksh-9999.ebuild [hunk 5]
```diff
@@ -12,29 +12,31 @@ if [[ $PV = 9999 ]]; then
...
 }
 
 src_install() {
```
Comment (mgorny) [inline app-shells/mksh/mksh-9999.ebuild]: I think the canonical way would be:

    into /
    dobin mksh

Other helpers shouldn't be affected by the prefix.

**Comment by mgorny** [comment]
Actually, I see it is already in baselayout-2.6, so please drop the shells updater.

File: app-shells/mksh/mksh-9999.ebuild (hunk 4 above)
Comment (lanodan) [reply app-shells/mksh/mksh-9999.ebuild]: It’s for Link Time Optimisation.
I didn’t notice earlier but upstream describes it as brittle so I’m removing it I guess.

**Comment by lanodan** [comment]
mksh is also present in baselayout-2.4.1 (which is still in the tree) btw, in case this would block removing it from the release ebuild.

File: app-shells/mksh/mksh-9999.ebuild [hunk 6]
```diff
@@ -29,14 +29,9 @@ S="${WORKDIR}/${PN}"
...
-	if use static
-	then
-		unset CC
-		export CC="/usr/bin/klcc"
-		export LDSTATIC="-static"
-	fi
+	if use static; then export CC="/usr/bin/klcc" export LDSTATIC="-static"; fi
 	export CPPFLAGS="${CPPFLAGS} -DMKSH_DEFAULT_PROFILEDIR=\\\"${EPREFIX}/etc\\\""
-	sh Build.sh -r -c lto || die
+	sh Build.sh -r || die
```
Comment (mgorny) [inline app-shells/mksh/mksh-9999.ebuild]: Can you make it respect `MAKEOPTS`, in particular `-j`?

File: app-shells/mksh/mksh-57-r1.ebuild [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+IUSE="static"
+DEPEND="static? ( dev-libs/klibc )"
+RDEPEND=""
+S="${WORKDIR}/${PN}"
+
+src_compile() {
+	tc-export CC
+	# we want to build static with klibc
+	if use static; then export CC="/usr/bin/klcc" export LDSTATIC="-static"; fi
+	export CPPFLAGS="${CPPFLAGS} -DMKSH_DEFAULT_PROFILEDIR=\\\"${EPREFIX}/etc\\\""
```
Comment (lanodan) [inline app-shells/mksh/mksh-57-r1.ebuild]: Oh, it would be better to use `append-cxxflags` from `flag-o-matic` instead, right?

**Comment by mgorny** [comment]
Please explore the possibiltiy of using `-M` option to `Build.sh` to split this into configure phase and use `emake` to do the actual build.

File: app-shells/mksh/mksh-57-r1.ebuild (hunk 7 above)
Comment (mgorny) [reply app-shells/mksh/mksh-57-r1.ebuild]: No, there's really no difference and no preference either way.

File: app-shells/mksh/mksh-9999.ebuild (hunk 6 above)
Comment (lanodan) [reply app-shells/mksh/mksh-9999.ebuild]: `-j` option is supported by their build script, would you have a recommendation on how to only get this option from `MAKEOPTS` as I’m pretty sure there is conflicting options?

Comment (mgorny) [reply app-shells/mksh/mksh-9999.ebuild]: Try playing with `-M` first. If that can't be gotten to generate something sane for make to use, multiprocessing.eclass has a helper to extract `-j`.

Comment (lanodan) [reply app-shells/mksh/mksh-9999.ebuild]: I tried to use `-M` but I failed to use it (and I didn’t find it in Void or NetBSD).
The `-j` flag actually doesn’t accepts any arguments and just allows a for-loop which doesn't wait for the command to finish, which might not be something that we want in gentoo.

**Comment by mgorny** [comment]
Could you squash those commits a bit? Or at least wait with making 57 from live ebuild till all fixes are in it.

**Comment by lanodan** [comment]
[2019-04-03 08:24:58+0000] Michał Górny:
> Could you squash those commits a bit? Or at least wait with making 57 from live ebuild till all fixes are in it.

Done, I put the 57-r1 bump as the last commit. Sorry for the delay.

**Comment by mgorny** [comment]
Have you tested it? Tests hang after `pass ./check.t:cd-pe` for me.

**Comment by lanodan** [comment]
[2019-04-07 01:24:20-0700] Michał Górny:
> Have you tested it? Tests hang after `pass ./check.t:cd-pe` for me.

Yes, both live and release runs fine with gcc or clang.

I checked at the content of check.t around cd-pe and it quite doesn’t
makes sense to hang on it, but well hanno@ wrote that the tests were
hanging too:

Subject: [gentoo-dev] Package up for grabs: app-shells/mksh
Message-ID: <20190113161219.75072865@computer>

I looked into b.g.o and mirabilos was saying that LTO could be the
issue, but well it didn’t cause me any with or without it.

It made me notice that this PR should probably close or at least be
mentionned in the following bugs:
https://bugs.gentoo.org/show_bug.cgi?id=536992
https://bugs.gentoo.org/show_bug.cgi?id=631948

Do I rebase or adding them in the Merge Commit is fine?

**Comment by mgorny** [comment]
Rebase, please. And if `-v` is indeed an option, please add it. We want failures as verbose as possible.

**Pull request closed without merging**